crates/nu-command/src/network/port.rs

```diff
@@ -121,25 +120,200 @@ fn get_free_port(
+        search_port_in_range((start_port..=end_port).into_spanned(range_span), call.head)?
+    };
+
+    Ok(Value::int(free_port as i64, call.head).into_pipeline_data())
+}
+
+fn system_provided_port() -> Result<u16, std::io::Error> {
+    TcpListener::bind("127.0.0.1:0")?
+        .local_addr()
+        .map(|addr| addr.port())
+}
+
+/// Find an open port by binding to every possible port in range.
+#[cfg(not(windows))]
+fn search_port_in_range(
+    range: Spanned<RangeInclusive<u16>>,
+    call_span: Span,
+) -> Result<u16, ShellError> {
+    use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+
+    let listener = 'search: {
+        let mut last_err = None;
+        for port in range.item {
+            let addr = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, port));
+            match TcpListener::bind(addr) {
+                Ok(listener) => break 'search Ok(listener),
+                Err(err) => last_err = Some(err),
             }
+        }
+
+        Err(IoError::new_with_additional_context(
+            last_err.expect("range not empty, validated before"),
+            range.span,
+            None,
+            "Every port has been tried, but no valid one was found",
+        ))
+    }?;
 
-            Err(IoError::new_with_additional_context(
-                last_err.expect("range not empty, validated before"),
-                range_span,
-                None,
-                "Every port has been tried, but no valid one was found",
-            ))
-        }?
+    Ok(listener
+        .local_addr()
+        .map_err(|err| IoError::new(err, call_span, None))?
+        .port())
+}
+
+#[cfg(windows)]
+mod windows {
+    use super::*;
+
+    use std::{
+        alloc::{Layout, alloc, dealloc},
+        ptr,
     };
 
-    let free_port = listener.local_addr().map_err(&from_io_error)?.port();
-    Ok(Value::int(free_port as i64, call.head).into_pipeline_data())
+    use ::windows::Win32::{
+        Foundation::{
+            ERROR_INSUFFICIENT_BUFFER, ERROR_INVALID_PARAMETER, ERROR_NOT_SUPPORTED, NO_ERROR,
+            WIN32_ERROR,
+        },
+        NetworkManagement::IpHelper::{GetTcpTable2, MIB_TCPROW2, MIB_TCPTABLE2},
+        Networking::WinSock::ntohs,
+    };
+
+    #[repr(C)]
+    struct TcpTable {
+        pub num_entries: u32,
+        pub table: [MIB_TCPROW2],
+    }
+
+    const _: () = assert!(align_of::<MIB_TCPTABLE2>() == 4);
+
+    impl TcpTable {
+        fn new() -> Result<Box<Self>, WIN32_ERROR> {
+            let mut size = 0;
+            let size_pointer: *mut u32 = &mut size;
+
+            // SAFETY:
+            // - Passing a null table pointer queries the required size (documented behavior).
+            // - `size_pointer` is a valid, non-null out pointer.
+            // - We expect `ERROR_INSUFFICIENT_BUFFER` so that `size` is written.
+            let ret_code = unsafe { GetTcpTable2(None, size_pointer, false) };
+            assert_eq!(WIN32_ERROR(ret_code), ERROR_INSUFFICIENT_BUFFER);
```

Comment:
My lawyer-y reading of https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-gettcptable2 is that it only guarantees that it will return the `ERROR_INSUFFICIENT_BUFFER` on passing of a NULL pointer to `(p)TcpTable`.